[JIT] Frozen Graph Conv-BN fusion

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50241 Re-enable no-op peepholes
* #50222 Add Post Freezing Optimizations, turn on by default in torch.jit.freeze
* #50221 Peephole Optimize out conv(x).dim(), which prevents BN fusion
* #50220 [JIT] Factor out peephole to own test file
* #50075 [JIT] Add Frozen Conv-> Add/Sub/Mul/Div fusion
* **#50074 [JIT] Frozen Graph Conv-BN fusion**

Adds Conv-BN fusion for models that have been frozen. I haven't explicitly tested perf yet but it should be equivalent to the results from @Chillee's PR [here](https://github.com/pytorch/pytorch/pull/476570) and [here](https://github.com/pytorch/pytorch/pull/47657#issuecomment-725752765). Click on the PR for details but it's a good speed up. 

 In a later PR in the stack I plan on making this optimization on by default as part of `torch.jit.freeze`. I will also in a later PR add a peephole so that there is not conv->batchnorm2d doesn't generate a conditional checking # dims.

Zino was working on freezing and left the team, so not really sure who should be reviewing this, but I dont care too much so long as I get a review 😛

Differential Revision: [D25856261](https://our.internmc.facebook.com/intern/diff/D25856261)

Commit:
Frozen Graph Conv-BN fusion

[ghstack-poisoned]

Commit:
Update on "Frozen Graph Conv-BN fusion"




[ghstack-poisoned]

Commit:
Update on "Frozen Graph Conv-BN fusion"




[ghstack-poisoned]

Commit:
Update on "Frozen Graph Conv-BN fusion"




[ghstack-poisoned]

ZolotukhinM commented on test/elias.py:
```diff
@@ -1,12 +0,0 @@
-import torch
-
-@torch.jit.script
-def foo(x):
-    return x + x + x
-
-torch._C._jit_override_can_fuse_on_cpu(True)
-
-foo(torch.rand([2], requires_grad=False))
-foo(torch.rand([2], requires_grad=False))
-foo(torch.rand([2], requires_grad=False))
-print(torch.jit.last_executed_optimized_graph())
```
Wait, is it in the trunk?😃

ZolotukhinM commented on torch/csrc/jit/passes/frozen_conv_folding.cpp:
```diff
@@ -0,0 +1,116 @@
+using Tensor = at::Tensor;
+
+bool nonConstantParameters(Node* n) {
+  for (size_t i = 1; i < n->inputs().size(); i++) {
```
Why do we start from 1 here?

eellison replied:
It was as of yesterday 😳

eellison replied:
oop, meant for the `test/elias.py` comment. We start from 1 because the input to conv isn't known, but we do need to know all other values (the conv/batchnorm parameeters)

ZolotukhinM commented on torch/csrc/jit/python/python_ir.cpp:
```diff
@@ -503,6 +503,17 @@ void initPythonIRBindings(PyObject* module_) {
           [](Node& n) {
             return py::make_iterator(n.inputs().begin(), n.inputs().end());
           })
+      .def(
+          "schema",
+          [](Node& n) {
+            std::stringstream ss;
+            if (auto sch = n.maybeSchema()) {
+              ss << n.schema();
+            } else {
+              ss << "(no schema)";
+            }
+            return ss.str();
+          })
```
Is it a leftover?

eellison replied:
No, I added this intentionally

ZolotukhinM commented on torch/csrc/jit/passes/frozen_conv_folding.cpp:
```diff
@@ -0,0 +1,116 @@
+  }
+}
+
+void FoldFrozenConvBatchnorm(Block* b) {
```
What is the difference between this pass and the one in `torch/csrc/jit/passes/fold_conv_bn.*`? Is it that this pass handles functional conv/bn sequences while the existing pass handles Conv-module/BN-module calls? If so, the name might be adjusted I think since it theoretically could occur in a non-frozen model too.

eellison replied:
The other pass handles modules. Basically, the other pass was written pre-freezing, I'm not sure there's any reason for it to exist anymore

ZolotukhinM replied:
Originally it was written for graph-mode quantization, which is replaced with an FX version now anyway. I think we should get rid of it then (cc @jerryzh168 ).

jerryzh168 replied:
this is added by @kimishpatel I think.

there are still people using torchscript graph mode quantization I think so we can't remove it right now.

eellison replied:
Doesn't graph mode quantization using freezing though? why can't we run fusion post - freezing?

jerryzh168 replied:
process for quantization is: fusion - prepare - convert - freeze
fusion needs to happen before freeze because it affects the weight of conv which will be observed in prepare and quantized in convert

kimishpatel replied:
As Jerry said the original conv+bn fusion cannot work post freezing. Plus since that one works at module levels we cannot quite make it work post freezing which inlines the graph. Ideally we add graph model fusion which works directly on graphs instead of modules, which this PR seems to be adding. Right now module based fusion is used in optimize_for_mobile. We can potentially move it after freezing. If this pass is for graph based fusion then is it necessary that it works only on frozen graph? Can it be made to work on non-frozen graphs too? That might be useful.

eellison replied:
There's not a great way to make this work on graphs that aren't frozen. You would have to remove the batchnorm weights from the parent module which isn't in general a valid transformation unless user-specified. That would also add a third conv + bn pass.

ZolotukhinM commented on test/jit/test_freezing.py:
```diff
@@ -1331,3 +1332,59 @@ def forward(self, x: torch.Tensor, key: str) -> Any:
+            self.run_pass("inline", scripted_mod.graph)
+
+            # TODO: add optimization for conv size
+            # Currrently a size check in batch norm gives conv multiple uses
+            # and prevents fusion
```
Does this mean that in practice this pass won't be able to do anything yet?

eellison replied:
Yes, for scripted models. I'm going to address that limitation later on in the stack

ZolotukhinM reviewed (approved):
Overall looks good, some comments are inline!

Commit:
Update on "[JIT] Frozen Graph Conv-BN fusion"


Adds Conv-BN fusion for models that have been frozen. I haven't explicitly tested perf yet but it should be equivalent to the results from @Chillee's PR [here](https://github.com/pytorch/pytorch/pull/476570) and [here](https://github.com/pytorch/pytorch/pull/47657#issuecomment-725752765). Click on the PR for details but it's a good speed up. 

 In a later PR in the stack I plan on making this optimization on by default as part of `torch.jit.freeze`. I will also in a later PR add a peephole so that there is not conv->batchnorm2d doesn't generate a conditional checking # dims.

Zino was working on freezing and left the team, so not really sure who should be reviewing this, but I dont care too much so long as I get a review 😛 





[ghstack-poisoned]

Commit:
Update on "[JIT] Frozen Graph Conv-BN fusion"


Adds Conv-BN fusion for models that have been frozen. I haven't explicitly tested perf yet but it should be equivalent to the results from @Chillee's PR [here](https://github.com/pytorch/pytorch/pull/476570) and [here](https://github.com/pytorch/pytorch/pull/47657#issuecomment-725752765). Click on the PR for details but it's a good speed up. 

 In a later PR in the stack I plan on making this optimization on by default as part of `torch.jit.freeze`. I will also in a later PR add a peephole so that there is not conv->batchnorm2d doesn't generate a conditional checking # dims.

Zino was working on freezing and left the team, so not really sure who should be reviewing this, but I dont care too much so long as I get a review 😛 





[ghstack-poisoned]

Commit:
Update on "[JIT] Frozen Graph Conv-BN fusion"


Adds Conv-BN fusion for models that have been frozen. I haven't explicitly tested perf yet but it should be equivalent to the results from @Chillee's PR [here](https://github.com/pytorch/pytorch/pull/476570) and [here](https://github.com/pytorch/pytorch/pull/47657#issuecomment-725752765). Click on the PR for details but it's a good speed up. 

 In a later PR in the stack I plan on making this optimization on by default as part of `torch.jit.freeze`. I will also in a later PR add a peephole so that there is not conv->batchnorm2d doesn't generate a conditional checking # dims.

Zino was working on freezing and left the team, so not really sure who should be reviewing this, but I dont care too much so long as I get a review 😛

Differential Revision: [D25856261](https://our.internmc.facebook.com/intern/diff/D25856261)

[ghstack-poisoned]
